### ISSUE TITLE
Long(ish) load times for QuPath project containing multiplexed images

### ISSUE THREAD

**Issue by MarkZaidi** [qupath/qupath#1154]
Currently building the latest version of QuPath; will close ticket if it ends up being fixed.

## Before we begin...

Before submitting your bug report, please check the following:

* [X] I've definitely found a bug (it you're not sure, please use [image.sc](https://forum.image.sc/tags/qupath) instead)
* [X] I've checked https://qupath.github.io for a new release that might already have fixed the issue
* [X] I've checked the [Changelog](https://github.com/qupath/qupath/blob/master/CHANGELOG.md) to see if the bug has already been fixed in the next release
* [X] I've checked for existing GitHub issues describing the same problem

## Bug report

**Describe the bug**
QuPath projects with several hundred IMC images take a notably longer time to load. Opening the images themselves is comparatively faster, yet are still read into memory. No noticeable change in memory usage is occurring when the QuPath project is being loaded, indicating that these images may not be pre-loaded (i.e. cached) during loading of the project itself. My guess is that the thumbnails are being reconstructed each time the QuPath project is loaded (computer must be restarted to reproduce, closing and relaunching QuPath is not sufficient). Given that thumbnail generation was updated in the changelog of the newest unofficial release, I'll build from source now and close the ticket if I can't reproduce the bug.

It's a minor inconvenience at best, especially since it's only a delay of two minutes. But keep in mind, the CPU here is a Ryzen 5950X with among the highest single thread speeds of desktop CPUs, and only 1 of the 32 threads is being used during this period. Depending on the root cause and a user's hardware configuration, load times may scale linearly or exponentially as project sizes move into the range of thousands of images.

If this is caused by thumbnail generation, then this may only be affecting multiplexed IHC images such as those acquired through IMC or CODEX.

**To Reproduce**
Video demonstration: https://www.youtube.com/watch?v=q4Jn9UTKUMw
1. Create a QuPath project
2. Load ~300 IMC images. Example IMC images can be found here: https://drive.google.com/file/d/1UcQmiIcjIfYdBVK1v6Phb32FBjHXciuH/view
3. Close QuPath and restart system
4. Launch QuPath and open the aforementioned project. Observe relatively long project load times

**Expected behavior**
Loading of a QuPath project should take a few seconds, especially if no measurements are present.

**Screenshots**
See video demonstration above. Note, memory usage doesn't change substantially when project is being loaded, however one thread is maxed at 100%, indicating that some kind of non-parallelized process is acting as the rate limiting step.

**Desktop (please complete the following information):**
 - OS: Windows 10
 - QuPath Version: 0.4.0 snapshot built on 2022-01-24

**Comment by MarkZaidi** [comment]
Closing as I can't seem to reproduce this issue anymore (yay!). Could have 'fixed' the project when launching in the current build, as noted by the changelog entry `Missing thumbnails are automatically regenerated when a project is opened`

**Comment by MarkZaidi** [comment]
Reopened as I still have the same issue with a newly built 0.4.0 (built on 2022-11-21). I can only seem to reproduce this once per day on average, as restarting the system will still yield fast load times. There's probably something being cached that gets wiped after a day, which substantially helps with project load times. My best guess is that it's the thumbnails, and since there isn't a way to disable them entirely, I can't confirm if that is the root cause.

Since only a single thread of the CPU is at 100% load when reproducing this bug, this will likely pose more of an issue in older computers and newer servers, since both will have low single-thread performance (takes ~4 mins to load the project on my system, and likely much longer on low single-thread systems).

**Comment by petebankhead** [comment]
If you use VisualVM then you should be able to find exactly what takes time.

Thumbnail generation can't safely be done in parallel because there's no guarantee the images will be pyramidal, so there is a high risk of out of memory errors.

That being said, I'm not certain thumbnail generation is the problem. If it's once per day, there could be some issue associated with update checking... but I can't see why that would be a problem and it wouldn't happen with the project is requested. VisualVM should resolve it one way or another in any case.

One other thing: is there anything conventional about where the project/images are stored? The need for a reboot makes me think there could be an operating system component. I've never seen it myself and don't recall anyone having reported it before.

**Comment by MarkZaidi** [comment]
>  is there anything conventional about where the project/images are stored?
The project is stored on the C drive, same physical SSD as the OS. The images are stored on the E drive, a separate 2TB SSD. Both have at least 300Gb of free space.

Will check out VisualVM and see if I can use it to monitor the initial project creation step.

**Comment by petebankhead** [comment]
Thanks, that would be good - it sounds like it would be difficult and time-consuming for anyone else to replicate the problem, and might not be possible if it turns out to be something more specific to your system. So solving this will likely require any any info you can get from VisualVM's CPU sampling.

I'll try to write up info about VisualVM + QuPath at some point, but there are brief instructions at https://forum.image.sc/t/designing-a-qupath-workstation/54849/14

QuPath shouldn't be regenerating thumbnails when a project is opened, and it shouldn't be remembering anything (other than the preferences) across relaunches.

v0.4.0 should generating missing thumbnails only, and do so in a background thread that doesn't block. Both v0.3 and v0.4 will try to access the thumbnail images in the project when it is opened; this *could* potentially block the UI, because it needs to be done in the UI thread (and there can only be one). However these should generally be small JPEGs so I've never known that to be a problem. Therefore I'd only expect this to be troublesome if there is some other reason why access to the disk is exceptionally slow.

I couldn't really tell what was going on in the video, partly because the analysis pane was closed - and when the image was opened I couldn't really see how or when that happened.

For slightly more info in the log, you can switch the log level in the preferences from `INFO` to `DEBUG`. Not sure if that will help narrow things down or not.

**Comment by MarkZaidi** [comment]
> So solving this will likely require any any info you can get from VisualVM's CPU sampling.

Since I wasn't sure which outputs to provide you with from VisualVM, I made a quick recording of the profiling as it ran while QuPath was loading the project: https://youtu.be/_GHn556qEAk

I wasn't able to get the profiler to work, despite installing a 64 bit version of Java 17, but the sampler and all other tools seemed to work. I've uploaded the thread and heap dumps to: https://drive.google.com/drive/folders/1akRolrnUwbryc3YMSRR_aXhJ5H6AgBWO?usp=sharing

> QuPath shouldn't be regenerating thumbnails when a project is opened, and it shouldn't be remembering anything (other than the preferences) across relaunches.
v0.4.0 should generating missing thumbnails only, and do so in a background thread that doesn't block. Both v0.3 and v0.4 will try to access the thumbnail images in the project when it is opened; this could potentially block the UI, because it needs to be done in the UI thread (and there can only be one). However these should generally be small JPEGs so I've never known that to be a problem. Therefore I'd only expect this to be troublesome if there is some other reason why access to the disk is exceptionally slow.

Based on my limited interpretation of the VisualVM results, I'd say that you're right. The lowest-level QuPath method I could find which was responsible for the ~4 minute load time is `qupath.lib.io.UriUpdater$SingleUriItem.getStatus()` which invokes `java.nio.file.File.exists()`

![image](https://user-images.githubusercontent.com/52012166/203870112-0ecd0822-7efd-4360-af2e-0858b8a03df7.png)

My guess is that there is some kind of checking going on, comparing the URI listed in the project, with the file path, to verify that each project entry is pointing to the correct image, and that's what's taking up CPU time.

> I couldn't really tell what was going on in the video, partly because the analysis pane was closed - and when the image was opened I couldn't really see how or when that happened.

In this recording, I've left the analysis pane open, and purposely triggered the "not responding" faded effect to show when the QuPath was loading the project. I've also opened a few demo images to show that read speeds weren't an issue.

Feel free to let me know if there's anything else I can provide to aid in resolving this issue.

**Comment by petebankhead** [comment]
Thanks, the screenshot suggests it's the `Files.exists()` check that is so slow. QuPath is probably calling this quite a lot, because it's usually really inexpensive - and *not* calling it leaves open the (admittedly small) possibility that the file has been deleted between calls.

I could try to cut down on this, but I'm skeptical that it will help because it sounds a lot like it's only the first call that is slow.

In fact, I'd be curious as to whether accessing each file is slow or just the first one on the disk. That should be relatively easy to test by creating a much smaller project and comparing the speed.

Either way, it sounds like the problem lurks in the realm of Java, the operating system, and the hard drive. I'm not sure if there's much can be done about it in QuPath.

I'd suggest:

* try accessing the images some other way before opening them in QuPath (e.g. through Windows Explorer), in case 'pre-warning' the OS is somehow useful
* try copying the images to a local drive (if possible)
* try timing a simple script with a file on the same disk, e.g.

```groovy
import java.nio.file.*
var path = '/Users/petebankhead/Desktop/Reduced middle button.png'
println "Exists: ${Files.exists(Paths.get(path))}"
```

**Comment by MarkZaidi** [comment]
> try accessing the images some other way before opening them in QuPath (e.g. through Windows Explorer), in case 'pre-warning' the OS is somehow useful

I think this is it! I opened the folder containing the images in explorer, and set the thumbnail size to "large icons" via `view > large icons`. This would generate thumbnails for all files that are currently displayed on screen. After those were generated, I scrolled down, so that the remaining files would also have their thumbnails generated.

Then, I launched the QuPath project in 0.4.1, and the load time was nearly instantaneous. Which, while resolving the issue, also means that the issue exists outside of QuPath.

So, for some reason, the thumbnail previews are lost after a period of time, and don't seem to coincide with closing the folder or restarting the computer. The first thing I attempted to preserve thumbnails was to ensure "Always show icons, never thumbnails" in the folder options was unchecked, however it already was by default:
![image](https://user-images.githubusercontent.com/52012166/211071879-ac70ef62-925f-4fe5-8ec3-10763d391393.png)
Toggling this field didn't seem to have any impact on project loading performance, since in either case, the thumbnails were already generated (just replaced with an icon if the box is checked).
Next, under Windows performance options, I noticed "save taskbar thumbnail previews" was unchecked by default. I've checked it, and so far, I don't seem to have trouble loading projects that already have thumbnails generated in explorer. I'll continue to test other projects and rebooting my computer, to confirm if this has resolved the issue.

TL;DR: **Check "save taskbar thumbnail previews" under Windows performance options. Subsequent loads of the project should now be much faster*

**Comment by MarkZaidi** [comment]
This seems to be fixed in 0.5.0-rc2, although I see no mention of this in the change log. Marking as closed.

**Comment by petebankhead** [comment]
Thanks for making it as closed. My best guess is that the issue didn't originate in QuPath, so was fixed as a side-effect of updating Java/other dependencies.
